### DISCUSSION TITLE
yarprunlogger doesn't have records & Driver <cartesiancontrollerclient> was found but could not open

### DISCUSSION THREAD

**Discussion by Yujia-Zhang0913** [robotology/community#663] [Q&A] [answered]
Hi, I use my own laptop to do the simulation of icub, but I have trouble viewing the log:
![Screenshot from 2023-11-01 11-21-16](https://github.com/robotology/community/assets/73634487/17c6a05a-713a-47c8-986f-b99ad9aa7f2f)
I wrote a code and named it myicub to use Icartesiancontrol, but the robot doesn't move. So I want to view the log, but I still cannot see log after I add yarplogger in the xml. Do I miss something?
And I saw from terminal that there is an error of myicub(when I changed main.cpp to the one in [tutorial_cartesian-interface](https://github.com/vvv-school/tutorial_cartesian-interface), the mistake is the same ).

The error is:

```console
[DEBUG] |yarp.dev.PolyDriver|cartesiancontrollerclient| Parameters are (device cartesiancontrollerclient) (local "/cartesian_client/left_arm") (remote "/icubSim/cartesianController/left_arm")
[INFO] |yarp.os.Port|/cartesian_client/left_arm/command:o| Port /cartesian_client/left_arm/command:o active at tcp://10.2.159.151:10094/
[INFO] |yarp.os.Port|/cartesian_client/left_arm/state:i| Port /cartesian_client/left_arm/state:i active at tcp://10.2.159.151:10095/
[INFO] |yarp.os.Port|/cartesian_client/left_arm/events:i| Port /cartesian_client/left_arm/events:i active at tcp://10.2.159.151:10096/
[INFO] |yarp.os.Port|/cartesian_client/left_arm/rpc:o| Port /cartesian_client/left_arm/rpc:o active at tcp://10.2.159.151:10097/
[INFO] |yarp.os.impl.PortCoreOutputUnit|/cartesian_client/left_arm/rpc:o| Sending output from /cartesian_client/left_arm/rpc:o to /icubSim/cartesianController/left_arm/rpc:i using tcp
[INFO] |yarp.os.impl.PortCoreOutputUnit|/cartesian_client/left_arm/command:o| Sending output from /cartesian_client/left_arm/command:o to /icubSim/cartesianController/left_arm/command:i using udp
[INFO] |yarp.os.impl.PortCoreInputUnit|/cartesian_client/left_arm/state:i| Receiving input from /icubSim/cartesianController/left_arm/state:o to /cartesian_client/left_arm/state:i using udp
[INFO] |yarp.os.impl.PortCoreInputUnit|/cartesian_client/left_arm/events:i| Receiving input from /icubSim/cartesianController/left_arm/events:o to /cartesian_client/left_arm/events:i using udp
[ERROR] unable to connect to solver!
[INFO] |yarp.os.impl.PortCoreOutputUnit|/cartesian_client/left_arm/command:o| output for route /cartesian_client/left_arm/command:o->udp->/icubSim/cartesianController/left_arm/command:i asking other side to close by out-of-band means
[INFO] |yarp.os.impl.PortCoreOutputUnit|/cartesian_client/left_arm/command:o| Removing output from /cartesian_client/left_arm/command:o to /icubSim/cartesianController/left_arm/command:i
[INFO] |yarp.os.impl.PortCoreInputUnit|/cartesian_client/left_arm/state:i| Removing input from /icubSim/cartesianController/left_arm/state:o to /cartesian_client/left_arm/state:i
[INFO] |yarp.os.impl.PortCoreInputUnit|/cartesian_client/left_arm/events:i| Removing input from /icubSim/cartesianController/left_arm/events:o to /cartesian_client/left_arm/events:i
[INFO] |yarp.os.impl.PortCoreOutputUnit|/cartesian_client/left_arm/rpc:o| Removing output from /cartesian_client/left_arm/rpc:o to /icubSim/cartesianController/left_arm/rpc:i
[ERROR] |yarp.dev.PolyDriver|cartesiancontrollerclient| Driver <cartesiancontrollerclient> was found but could not open
```

I cannot figure out why. Really thanks!

**Comment by Yujia-Zhang0913** [comment]
I have figured out the question about yarplogger.
Now the log is like this:
[log_icubsrv_iKinCartesianSolver_18636.log](https://github.com/robotology/community/files/13235723/log_icubsrv_iKinCartesianSolver_18636.log)
[log_icubsrv_yarprobotinterface_15945.log](https://github.com/robotology/community/files/13236157/log_icubsrv_yarprobotinterface_15945.log)
But I think in my xml, it is icubsim not icub?
![Screenshot from 2023-11-02 09-28-55](https://github.com/robotology/community/assets/73634487/021e86b5-11b3-48a7-b8cc-9b759aa6acc5)

Really thanks

**Reply by pattacini** [reply]
Hi @Yujia-Zhang0913 

From the log, I spotted the following message:
```console
<WARNING> Cannot open device torso
```
indicating that the torso part is not available. This part is essential to run all the Cartesian Control stack.

Just try to launch Gazebo only and see whether torso is active:
```console
yarp name list | grep -i torso
```

If it is not active, it may be because:
- an address conflict;
- a wrong Gazebo model.

Then, from the first log, it appears that the solver isn't configured to run alongside the simulator but with the physical robot, instead. This is probably a matter of configuration files.

**Reply by Yujia-Zhang0913** [reply]
Hi, @pattacini 
I try to launch my code today, and somehow the errors do not repeat(I guess maybe it is because an address conflict, since I restarted the computer)
But my main.cpp still have some problem running
[log_gaga-WRTB-WXX9_myicub_9982.log](https://github.com/robotology/community/files/13246104/log_gaga-WRTB-WXX9_myicub_9982.log)
and my code is
```cpp
// -*- mode:C++; tab-width:4; c-basic-offset:4; indent-tabs-mode:nil -*-

#include <string>
#include <cstdio>

#include <yarp/os/Network.h>
#include <yarp/dev/ControlBoardInterfaces.h>
#include <yarp/dev/PolyDriver.h>
#include <yarp/os/Time.h>
#include <yarp/sig/Vector.h>
#include <iCub/iKin/iKinFwd.h>
#include <yarp/dev/CartesianControl.h>
#include <yarp/dev/GazeControl.h>
using namespace std;
using namespace yarp::dev;
using namespace yarp::sig;
using namespace yarp::os;

int main(int argc, char *argv[]) 
{
    Network yarp;
  
    Property params;
    params.fromCommand(argc, argv);
    if (!params.check("robot"))
    {
        fprintf(stderr, "Please specify the name of the robot\n");
        fprintf(stderr, "--robot name (e.g. icub)\n");
        return 1;
    }
    std::string robotName=params.find("robot").asString();
    std::string remotePorts="/";
    remotePorts+=robotName;
    remotePorts+="/right_arm";

    std::string localPorts="/test/client";
    Property options;
    options.put("device", "remote_controlboard");
    options.put("remote", robotName+"/right_arm");
    options.put("local", "/test/client");
    IEncoders *iencs;


    PolyDriver driver_arm;
    driver_arm.open(options);
    driver_arm.view(iencs);

    Property option;
    option.put("device","cartesiancontrollerclient");
    option.put("remote",robotName+"/cartesianController/right_arm");
    option.put("local","/client/right_arm");

    PolyDriver clientCartCtrl(option);

    ICartesianControl *icart=NULL;

    if (clientCartCtrl.isValid()) {
    clientCartCtrl.view(icart);
    }

    int times=0;
    Vector x0,o0;
    icart->getPose(x0,o0);
    printf("%.1lf %.1lf %.1lf\n", x0[0], x0[1], x0[2]);
    Vector xd0=x0; xd0[1]+=-0.1;
    Vector od0=o0;
    icart->goToPose(xd0,od0); 
    Vector xd,od;
    icart->goToPoseSync(xd0,od0);
    printf("%.1lf %.1lf %.1lf\n", xd0[0], xd0[1], xd0[2]);
    while(true)
    {
        if (times%2)
        {
            xd=xd0;
            od=od0;
        }
        else
        {
            xd=x0;
            od=o0;
        }
        times++;
        icart->goToPose(xd,od);   // send request and wait for reply

        bool done=false;
        while (!done) {
        icart->checkMotionDone(&done);
        Time::delay(0.04);   // or any suitable delay
        }
    }
clientCartCtrl.close();
driver_arm.close();
return 0;
}
```

And when I use the main.cpp in tutorial
the log is like
[log_gaga-WRTB-WXX9_myicub_14014.log](https://github.com/robotology/community/files/13246177/log_gaga-WRTB-WXX9_myicub_14014.log)

And one more question, does icub has self-collision detection? How can I detect whether it is collided? Thank you.

**Reply by pattacini** [reply, marked as answer]
Hi @Yujia-Zhang0913 

Don't know what exact tutorial you're referring to.

At any rate, from the first log, it seems that you're trying to connect to `icubSim/right_arm` and `icubSim/cartesianController/right_arm`, whereas YARP ports shall begin with the leading forward slash `/`.

Therefore, correct names for destination ports are instead:
- `/icubSim/right_arm`
- `/icubSim/cartesianController/right_arm`

>And one more question, does icub has self-collision detection? How can I detect whether it is collided? Thank you.

Generally speaking, the Cartesian Control exploits only a limited knowledge for doing self-collision avoidance. In particular, the shoulder and the elbow are controlled in such a way that they won't hit the trunk.

That said, it's definitely possible to command target poses that will make the hand bump against the body.
However, simple/naive yet effective collision avoidance in Cartesian space shouldn't be that difficult to implement if one resorts to cylindrical coordinates. More accurate algorithms can be very task dependent though.

**Reply by Yujia-Zhang0913** [reply]
Hi @pattacini 
Really thanks for helping me. Now icub works perfectly!

**Reply by pattacini** [reply]
Happy to know that!
